SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
+// An IndexSpaceRegistry is a context specific, modifiable, list of registered indices.
+// It is the task of the USER to ensure that all spaces for their context are included.
+// the result of certain set operations such as intersection and unIon require access to the registry
+//
+class IndexSpaceRegistry{
+ public:
+  IndexSpaceRegistry(){
+    //register nulltype
+    this->add(nulltype);
   }
 
-  /// @param key the key associated with a registered IndexSpace; this can be
-  /// either
-  ///            the base key used to invoke `IndexSpace::register_instance()`
-  ///            or a key used to invoke `IndexSpace::register_key()`
-  /// @return the attribute of the IndexSpace object corresponding to @p key
-  /// @throw bad_key if \p key has not been registered
-  static Attr to_attr(std::wstring_view key) {
-    const auto found_it = key2attr_.find(key);
-    if (found_it != key2attr_.end()) return found_it->second;
-    throw bad_key();
+  //retrieve an IndexSpace from the registry by the label
+  IndexSpace retrieve(std::wstring_view label) const{
+    for(auto it = label_space.begin(); it != label_space.end(); ++it){
+      if(IndexSpace::reduce_key(label) == it->first){
+        return it->second;
+      }
+    }
+    throw std::invalid_argument("The label you are trying to retrieve has not been registered");
+    return label_space.at(L"");
   }
 
-  static std::wstring to_wstring(std::wstring_view key) {
-    return std::wstring(key.begin(), key.end());
+  //add an IndexSpace to this registry. duplicate type bitsets forbidden
+  void add(IndexSpace IS){
+    if(!(find_IndexSpace(IS) == nulltype) ){
+      throw std::invalid_argument("The registry already has a TypeAttr(bitset) corresponding to the IndexSpace you are trying to add! "
+          "If you are trying to replace the index use the replace(IndexSpace) function");
+    }
+    else {
+      label_space.emplace(IS.get_base_key(), IS);
+    }
+  }
+
+  // return a vector of pairs since we will be accessing by index not key.
+  std::vector<std::pair<IndexSpace,std::wstring_view>> base_spaces_label(){
+    std::vector<std::pair<IndexSpace,std::wstring_view>> result;
+    //if the registered instance has a single bit true, it is a base space.
+    // std::map should order this new map via IndexSpace.
+    ///TODO check that this is ordered correctly
+    for (auto begin = label_space.begin(); begin != label_space.end(); begin++){
+      if(has_single_bit(begin->second.type().to_int32())){
+        result.push_back({begin->second,begin->first});
+      }
+    }
+    std::sort(result.begin(),result.end(),
+    [](std::pair<IndexSpace,std::wstring_view> a, std::pair<IndexSpace,std::wstring_view> b)-> bool{
+      return a.first.type() < b.first.type();
+    });
+    return result;
+  }
+
+  // clear the label_space map essentially clearing the registry
+  void clear_registry(){
+    label_space.clear();
+    IndexSpace nulltype(L"",0,0);
+    this->add(nulltype);
```

Comment:
```suggestion
    this->add(nulltype);
```
Presumably, we want to reuse the class-level `nulltype` instance